### DISCUSSION TITLE
Initialise Actor from DB best practice

### DISCUSSION THREAD

**Discussion by mika76** [akkadotnet/akka.net#4777] [Q&A]
I'm wondering what's the best way to init an actor which needs some data from a db before it starts working.

I had assumed just put the db calls into the PreStart() method but I notice it's not async and I also do not see much info about what happens if an exception is thrown in the PreStart() method. So basically I'm wondering if there is a better/more correct way to do this simple thing?

**Comment by mika76** [comment]
Ok to update here, I chose to go with an `Initialize` message which the actor sends to itself in the `PreStart()` method. Seems to work ok, but I'd love to know if there are better ways maybe...

EDIT: Would also love to know what happens on restart? What if there's messages already in the mailbox...

**Reply by coldwarrl** [reply]
Have you see https://getakka.net/articles/actors/receive-actor-api.html#initialization-via-prestart ?

I would also send an init message itself to the actor. My usual pattern is to have the state machine, mentioned in the link above, that ensures that the actor does not handle post init messages (work requests from other actors) if it is not initialized, yet. If other actors need to be aware of the state change, one option is to broadcast this to the event bus (cluster pub/sub or local event bus) for instance. I do that in some actors, which have a quite complex state model and so the whole system is aware of key states in the system.

**Reply by mika76** [reply]
Thanks @coldwarrl - so I guess I'm now following the suggested init message method stated in https://getakka.net/articles/actors/receive-actor-api.html#initialization-via-message-passing but there is a big warning under there which made me uneasy. My actors are pretty simple and local though so I'm not expecting any problems for the moment.

I do wonder though if PreStart shouldn't be an async method returning a Task, I can definitely see specific initialisations that might take a while but are necessary for an actor to start.
